File: edge-app/core/internal/agent/neutral.go
package agent

// Der gefuehrte Neutral-Zeit-Test (docs/ota-autonomie.md §3): die Kalibrierung
// der Inverter-Neutral-Zeit T am GERAET statt am Pruefstand.
//
// Bisher war T nur mit einem physisch gezogenen Kabel zu belegen
// (VP_OTA_NEUTRAL_VERIFIED). Dieser Weg macht T zu einer MESSBAREN
// Eigenschaft, die die Box mit ihrem eigenen Lesepfad selbst ermittelt: ein
// kleiner, klar von neutral abweichender Sollwert wird geschrieben und
// bestaetigt (Register-Rueckmeldung), danach hoert die Box bewusst auf, ihn
// aufzufrischen, und beobachtet ueber den GEWOEHNLICHEN Telemetrie-Lesepfad -
// der unveraendert weiterlaeuft -, wie lange der Wechselrichter ihn haelt,
// bis er von selbst neutral wird. Das volle Verfahren + die Sicherheits-
// Argumentation stehen in internal/neutralcal (rein, ohne I/O).
//
// Diese Datei verdrahtet den Wechselrichter-Wert (`neutralTestOverride`, auf
// edge/setpoint - oder eben NICHTS, siehe unten), die Beobachtung aus der
// normalen Telemetrie (`neutralObserve`, aufgerufen aus onLocalTelemetry) und
// die Register-Bestaetigung aus dem Steuer-Readback (onControlReadback in
// agent.go). Die Freigabe folgt EXAKT demselben Muster wie First-Light
// Kalibrierung + PV-Abregelung:
//
//   - derselbe globale Not-Aus (VP_CONTROL_ENABLED) gilt unveraendert;
//   - der Testwert laeuft durch DIESELBE guards.Clamp-Kette wie jeder andere
//     Sollwert - nie um sie herum;
//   - eine harte Obergrenze (neutralcal.DefaultTTL) plus eine engere Grenze
//     fuer die Schreibphase (neutralcal.DepartureTimeout) begrenzen die
//     Messdauer, und ein `time.AfterFunc`-Wachhund (unabhaengig vom Takt der
//     Oberflaeche) nimmt die Steuerung spaetestens dann wieder auf;
//   - jederzeit per Knopf abbrechbar (NeutralAbort), und ein Abbruch
//     entwertet den Nachweis sofort;
//   - ein Sicherheitsnetz bricht den Test proaktiv ab, sobald etwas ANDERES
//     die Anlage braucht: eine laufende Kalibrierung, oder eine EILIGE OTA-
//     Aktualisierung, die die Anlage neutral parken moechte (die
//     otaNeutralOverride-Bitte darf nie fuer die Dauer eines Neutral-Zeit-
//     Tests verhungern);
//   - der Test AENDERT NIE eine Guard-Grenze und weitet NIE eine Befugnis -
//     ein bestandener Test traegt ausschliesslich eine gemessene
//     Sekundenzahl in die geraete-lokale Beleg-Datei ein
//     (otaapply.SaveNeutralRecord), die otaapply.NeutralTable.ForWithMeasured
//     GLEICHWERTIG zu VP_OTA_NEUTRAL_VERIFIED heranzieht - mit Vorrang fuer
//     die Umgebungsvariable.

import (
	"encoding/json"
	"fmt"
	"log/slog"
	"strings"
	"time"

	"git.tecmaxx.de/mamotec/voltpilot-ems/edge-app/core/internal/cloud"
	"git.tecmaxx.de/mamotec/voltpilot-ems/edge-app/core/internal/guards"
	"git.tecmaxx.de/mamotec/voltpilot-ems/edge-app/core/internal/localbus"
	"git.tecmaxx.de/mamotec/voltpilot-ems/edge-app/core/internal/neutralcal"
	"git.tecmaxx.de/mamotec/voltpilot-ems/edge-app/core/internal/otaapply"
	"git.tecmaxx.de/mamotec/voltpilot-ems/edge-app/core/internal/state"
)

// neutralErr builds a user-facing (German) neutralcal.ValidationError the web
// layer maps to HTTP 400 (mirrors calErr/curtailErr).
func neutralErr(format string, args ...any) error {
	return &neutralcal.ValidationError{Msg: fmt.Sprintf(format, args...)}
}

// neutralPreflight is the shared guard for starting a test: the global
// kill-switch must be on and a battery-controllable inverter must be
// selected - the SAME preflight calibration uses, since the test writes
// through the same bounded-write mechanism.
func (a *Agent) neutralPreflight() error {
	if !a.Cfg.ControlEnabled {
		return neutralErr("Die Wechselrichter-Steuerung ist deaktiviert (Not-Aus). Neutral-Zeit-Test nicht möglich.")
	}
	sel, ok := a.GetInverter()
	if !ok {
		return neutralErr("Bitte wählen Sie zuerst Ihren Wechselrichter aus.")
	}
	if !calibrationControllable(sel) {
		return neutralErr("Für diesen Wechselrichter ist kein Neutral-Zeit-Test verfügbar.")
	}
	a.calMu.Lock()
	engaged := a.cal.Engaged(time.Now().UTC())
	a.calMu.Unlock()
	if engaged {
		return neutralErr("Es läuft gerade eine Kalibrierung - bitte zuerst beenden.")
	}
	return nil
}

// NeutralSnapshot implements web.CalibrationController (GET /api/neutral).
func (a *Agent) NeutralSnapshot() neutralcal.View {
	return a.neutralView(time.Now().UTC())
}

// NeutralStartTest implements web.CalibrationController's neutral-time entry
// point (folded into the SAME calGuard-protected surface).
func (a *Agent) NeutralStartTest() (neutralcal.View, error) {
	now := time.Now().UTC()
	if err := a.neutralPreflight(); err != nil {
		return a.neutralView(now), err
	}
	family := a.currentFamily()
	a.neutralMu.Lock()
	testKw, err := a.neutralCal.Start(family, now)
	if err == nil {
		if a.neutralWatchdog != nil {
			a.neutralWatchdog.Stop()
		}
		// Belt-and-suspenders: independent of the setpoint tick cadence, this
		// resumes normal control shortly after the hard TTL - the same
		// pattern as calibration's/curtailment's own watchdog.
		a.neutralWatchdog = time.AfterFunc(a.neutralCal.TTL()+2*time.Second, a.nudgeSetpoint)
	}
	a.neutralMu.Unlock()
	if err != nil {
		return a.neutralView(now), err
	}
	slog.Warn("Neutral-Zeit-Test gestartet", "family", family, "test_kw", testKw)
	a.nudgeSetpoint() // publish the tiny departure immediately
	return a.neutralView(now), nil
}

// NeutralAbort ends the running test immediately - control resumes on the
// very next setpoint tick, and the evidence is invalidated (an aborted run
// proves nothing).
func (a *Agent) NeutralAbort() neutralcal.View {
	now := time.Now().UTC()
	a.neutralMu.Lock()
	a.neutralCal.Abort(now)
	if a.neutralWatchdog != nil {
		a.neutralWatchdog.Stop()
	}
	a.neutralMu.Unlock()
	a.nudgeSetpoint()
	return a.neutralView(now)
}

// NeutralRecord is "Als Nachweis übernehmen": persists the current, valid,
// PASSED evidence into ota/neutral-verified.json for this family. Refused
// (400) unless the evidence is a genuine, still-valid pass with a measured
// duration - a "nicht beweisbar"/"kein_nachweis" run can NEVER be recorded,
// by construction (CanRecord mirrors that gate).
func (a *Agent) NeutralRecord() (neutralcal.View, error) {
	now := time.Now().UTC()
	family := a.currentFamily()
	if family == "" {
		return a.neutralView(now), neutralErr("Kein Wechselrichter ausgewählt.")
	}
	a.neutralMu.Lock()
	canRecord := a.neutralCal.CanRecord(family, now)
	var ev *neutralcal.Evidence
	if canRecord {
		ev = a.neutralCal.EvidenceFor(family, now)
	}
	a.neutralMu.Unlock()
	if !canRecord || ev == nil || ev.MeasuredSeconds == nil {
		return a.neutralView(now), neutralErr("Es liegt kein bestandener, noch gültiger Nachweis vor - bitte zuerst einen Test durchführen.")
	}
	rec := otaapply.NeutralRecord{
		Family:        family,
		Seconds:       *ev.MeasuredSeconds,
		MeasuredAt:    now.Format(otaapply.TimeFormat),
		TestKw:        ev.TestKw,
		SettleSamples: ev.SettleSamples,
	}
	if err := otaapply.SaveNeutralRecord(a.Cfg.DataDir, rec); err != nil {
		return a.neutralView(now), neutralErr("Der Nachweis konnte nicht gespeichert werden: %s", err.Error())
	}
	slog.Warn("Neutral-Zeit-Test: belegtes T eingetragen", "family", family, "seconds", rec.Seconds)
	return a.neutralView(now), nil
}

// neutralObserve feeds a fresh measured battery-power reading into a
// running test - called from onLocalTelemetry with EVERY sample, exactly the
// "normal read path keeps running unchanged" the mechanism relies on.
func (a *Agent) neutralObserve(batteryKw float64, now time.Time) {
	family := a.currentFamily()
	if family == "" {
		return
	}
	a.neutralMu.Lock()
	a.neutralCal.Observe(family, batteryKw, now)
	a.neutralMu.Unlock()
}

// neutralTestOverride handles the ENTIRE setpoint tick while a Neutral-Zeit-
// Test is running: either publishing the tiny bounded departure (still
// guard-clamped, still gated by the global kill-switch) or, once the
// departure is confirmed, publishing NOTHING AT ALL - the whole mechanism.
// Returns true when it handled the tick (applySetpoint must return early).
func (a *Agent) neutralTestOverride(now time.Time, r guards.Reading, limits guards.Limits) bool {
	family := a.currentFamily()

	a.neutralMu.Lock()
	if !a.neutralCal.Active(now) {
		a.neutralMu.Unlock()
		return false
	}
	// Sicherheitsnetz "Interlock": eine EILIGE OTA-Aktualisierung darf fuer
	// die Dauer eines Neutral-Zeit-Tests (bis zu neutralcal.DefaultTTL) nicht
	// verhungern - otaNeutralOverride sitzt in der Kette NACH diesem
	// Aufrufer und kaeme sonst nie zum Zug. Der Test bricht dann ab; er hat
	// ohnehin nichts mehr zu beweisen, sobald etwas anderes die Anlage
	// braucht.
	if a.otaNeutralRequestPending(now) || family == "" || family != a.neutralCal.ActiveFamily(now) {
		a.neutralCal.Abort(now)
		a.neutralMu.Unlock()
		if a.neutralWatchdog != nil {
			a.neutralWatchdog.Stop()
		}
		slog.Warn("Neutral-Zeit-Test abgebrochen (Interlock): die Anlage wird anderweitig gebraucht")
		return false
	}
	kw, publish := a.neutralCal.Command(family, now)
	a.neutralMu.Unlock()

	if !publish {
		// SILENT phase (or the test just concluded): publish NOTHING - going
		// completely silent on edge/setpoint is the entire measurement
		// mechanism (see internal/neutralcal's package doc). The mode is
		// still reported honestly so the card/heartbeat never look like a
		// stuck setpoint.
		a.State.Update(func(s *state.Snapshot) { s.Mode = state.ModeNeutralTest })
		return true
	}

	kwClamped := guards.Clamp(kw, limits, r)
	certified := a.controlCertified(family)
	msg := map[string]any{
		"battery_setpoint_kw": kwClamped,
		"source":              "neutral_test",
		"ts":                  now.Format(time.RFC3339Nano),
		"control_enabled":     a.Cfg.ControlEnabled,
		"device_certified":    certified,
		// The executor's certification bypass marker - the SAME one
		// calibration uses, so an as-yet-uncertified family can still prove
		// its neutral-time here (measuring T is a prerequisite for autonomous
		// OTA, not a consequence of it).
		"calibration":         true,
		"grid_charge_allowed": false,
		"soc_min_pct":         a.Cfg.SocMinPct,
		"soc_max_pct":         a.Cfg.SocMaxPct,
	}
	if certified {
		if p := a.certifiedControlPath(family); p != "" {
			msg["device_certified_path"] = p
		}
	}
	if cur := a.curtailSetpointExtras(now); cur != nil {
		msg["curtail"] = cur
	}
	raw, _ := json.Marshal(msg)
	if err := a.Bus.Publish(localbus.TopicSetpoint, raw, true); err != nil {
		slog.Error("neutral-time test setpoint publish failed", "err", err)
	}
	a.neutralMu.Lock()
	a.neutralCal.NoteWrite(family, now)
	a.neutralMu.Unlock()
	a.State.Update(func(s *state.Snapshot) {
		s.Mode = state.ModeNeutralTest
		s.SetpointKw = kwClamped
		s.SlotStart = time.Time{}
		s.ControlEnabled = a.Cfg.ControlEnabled
		s.ControlCertified = certified
		s.PeakGuardActive = false
		s.PeakQuarterMeanKw = nil
		s.Trim = nil
		s.Follow = nil
	})
	a.trim.Release()
	a.follow.Release()
	return true
}

// neutralView builds GET /api/neutral: session state + a persisted-record
// echo for the currently selected family.
func (a *Agent) neutralView(now time.Time) neutralcal.View {
	family := a.currentFamily()
	v := neutralcal.View{
		ControlEnabled:          a.Cfg.ControlEnabled,
		AdminGate:               a.Cfg.CalibrationAdminSecret != "",
		Family:                  family,
		TestKw:                  neutralcal.TestKw,
		TestTTLSeconds:          int(a.neutralTTL() / time.Second),
		DepartureTimeoutSeconds: int(neutralcal.DepartureTimeout / time.Second),
	}
	if b := a.liveBattKw(); b != nil {
		v.LiveBatteryKw = b
	}
	if family == "" {
		v.Reason = "Bitte wählen Sie zuerst Ihren Wechselrichter aus."
		return v
	}
	sel, ok := a.GetInverter()
	if !ok || !calibrationControllable(sel) {
		v.Reason = "Für diesen Wechselrichter ist kein Neutral-Zeit-Test verfügbar."
		return v
	}
	v.Available = true

	a.neutralMu.Lock()
	phase := a.neutralCal.Phase(family, now)
	active := phase == neutralcal.PhaseActive || phase == neutralcal.PhaseSilent
	var (
		testKw float64
		pub    bool
	)
	if active {
		testKw, pub = a.neutralCal.Command(family, now)
	}
	ev := a.neutralCal.EvidenceFor(family, now)
	canRecord := a.neutralCal.CanRecord(family, now)
	a.neutralMu.Unlock()

	if active {
		tv := &neutralcal.TestView{
			Family: family,
			TestKw: neutralcal.TestKw,
			Phase:  phase,
		}
		if pub {
			tv.TestKw = testKw
		}
		if ev != nil {
			tv.RegisterConfirmed = ev.RegisterConfirmed
			tv.DepartureConfirmed = ev.DepartureConfirmed
			tv.SettleSamples = ev.SettleSamples
			tv.SettleRequired = ev.SettleRequired
		}
		a.neutralMu.Lock()
		tv.SecondsRemaining = a.neutralCal.RemainingSeconds(family, now)
		a.neutralMu.Unlock()
		if b := a.liveBattKw(); b != nil {
			tv.LiveBatteryKw = b
		}
		v.Test = tv
	}
	v.Evidence = ev
	v.CanRecord = canRecord

	if recs := otaapply.LoadNeutralEvidence(a.Cfg.DataDir); recs != nil {
		if rec, ok := recs[strings.ToLower(strings.TrimSpace(family))]; ok {
			v.Recorded = &neutralcal.RecordedView{Seconds: rec.Seconds, MeasuredAt: rec.MeasuredAt}
		}
	}
	return v
}

func (a *Agent) neutralTTL() time.Duration {
	if a.neutralCal == nil {
		return neutralcal.DefaultTTL
	}
	return a.neutralCal.TTL()
}

// liveBattKw reads the latest measured battery power (nil = none yet).
func (a *Agent) liveBattKw() *float64 {
	a.mu.Lock()
	defer a.mu.Unlock()
	if a.lastBattKw == nil {
		return nil
	}
	v := *a.lastBattKw
	return &v
}

// neutralVerifiedSummary is the additive heartbeat fact: the device-measured
// Neutral-Zeit T for the CURRENTLY selected family, if this device ever
// recorded one. It is a FACT, never an authorization - whether it actually
// opens the autonomous-apply gate is decided entirely on-device by
// otaapply.NeutralTable.ForWithMeasured (VP_OTA_NEUTRAL_VERIFIED still wins).
// nil = never measured here (or no inverter selected) - an older backend
// simply ignores the field.
func (a *Agent) neutralVerifiedSummary() *cloud.NeutralVerifiedSummary {
	family := strings.ToLower(strings.TrimSpace(a.currentFamily()))
	if family == "" {
		return nil
	}
	recs := otaapply.LoadNeutralEvidence(a.Cfg.DataDir)
	rec, ok := recs[family]
	if !ok {
		return nil
	}
	return &cloud.NeutralVerifiedSummary{
		Family: rec.Family, Seconds: rec.Seconds, MeasuredAt: rec.MeasuredAt,
	}
}

File: edge-app/core/internal/web/static/neutral.js
// VoltPilot Edge - Neutral-Zeit messen (docs/ota-autonomie.md §3).
// The guided First-Light test that MEASURES the inverter's
// Kommunikations-Verlust-Zeit T at the device instead of a bench session
// with a physically cut cable: command a small departure from neutral,
// confirm it landed, then go completely silent while the normal telemetry
// read path watches the return. Mutations share the calibration admin gate
// (X-VP-Calibration-Token, sessionStorage "vp.cal.token" - the same token
// calibration.js/curtail.js already prompt for on this device). Polls
// GET /api/neutral while the card is on screen.
(function (global) {
  "use strict";

  var $ = function (id) { return global.document.getElementById(id); };

  var adminToken = "";
  try { adminToken = sessionStorage.getItem("vp.cal.token") || ""; } catch (e) { adminToken = ""; }

  function post(path, body) {
    var headers = { "Content-Type": "application/json" };
    if (adminToken) headers["X-VP-Calibration-Token"] = adminToken;
    return fetch(path, {
      method: "POST",
      headers: headers,
      body: body === undefined ? undefined : JSON.stringify(body),
    }).then(function (r) {
      return r.json().then(function (j) { return { ok: r.ok, status: r.status, body: j }; },
        function () { return { ok: r.ok, status: r.status, body: {} }; });
    });
  }

  function showErr(msg) {
    var e = $("neutralErr");
    if (!e) return;
    e.textContent = msg || "";
    e.hidden = !msg;
  }

  // A 401 means the admin gate rejected the token: prompt once (tab-scoped),
  // exactly like the battery calibration + PV-Abregelung cards.
  function ensureToken(resp) {
    if (resp.status !== 401 && !(resp.body && resp.body.auth_required)) return false;
    var val = "";
    try { val = global.prompt("Die Kalibrierung ist geschützt. Administrator-Kennwort:") || ""; } catch (e) { val = ""; }
    if (!val) { showErr("Ohne Administrator-Kennwort sind keine Änderungen möglich."); return false; }
    adminToken = val;
    try { sessionStorage.setItem("vp.cal.token", val); } catch (e) { /* tab-only */ }
    return true; // caller may retry once
  }

  function act(path, body) {
    showErr("");
    post(path, body).then(function (resp) {
      if (ensureToken(resp)) {
        post(path, body).then(function (retry) {
          if (!retry.ok) showErr((retry.body && retry.body.error) || "Aktion fehlgeschlagen.");
          refresh();
        });
        return;
      }
      if (!resp.ok) showErr((resp.body && resp.body.error) || "Aktion fehlgeschlagen.");
      refresh();
    });
  }

  function pill(text, tone) {
    return '<span class="pill ' + (tone || "") + '">' + text + "</span>";
  }

  // The honest verdict, in the operator's words - never disguise "nicht
  // beweisbar" (a statement about the RUN, not a defect of the inverter) as
  // either a pass or a failure.
  function verdictPill(verdict) {
    switch (verdict) {
      case "bestanden": return pill("bestanden", "ok");
      case "nicht_beweisbar": return pill("nicht beweisbar", "warn");
      case "kein_nachweis": return pill("kein Nachweis", "warn");
      case "laeuft": return pill("läuft …", "");
      default: return "";
    }
  }

  function render(v) {
    var card = $("neutralCard");
    if (!card) return;
    card.hidden = !v;
    if (!v) return;

    var empty = $("neutralEmpty");
    var body = $("neutralBody");
    if (!body) return;
    if (!v.available) {
      if (empty) { empty.hidden = false; var t = $("neutralEmptyText"); if (t) t.textContent = v.reason || "Neutral-Zeit-Test derzeit nicht verfügbar."; }
      body.hidden = true;
      return;
    }
    if (empty) empty.hidden = true;
    body.hidden = false;

    var lines = [];
    var actions = "";

    if (v.recorded) {
      lines.push("Belegte Neutral-Zeit: " + v.recorded.seconds + " s (gemessen " + v.recorded.measured_at + ")");
    } else {
      lines.push("Für dieses Modell ist noch keine Neutral-Zeit belegt.");
    }

    if (v.test) {
      var phaseText = v.test.phase === "aktiv"
        ? "Test läuft: Sollwert " + v.test.test_kw + " kW wird geschrieben und bestätigt …"
        : v.test.phase === "still"
          ? "Test läuft: es wird JETZT NICHTS mehr geschrieben - die Box beobachtet, wie lange der Wechselrichter den Sollwert von selbst hält."
          : "Test läuft …";
      lines.push(phaseText);
      lines.push("Register: " + (v.test.register_confirmed ? "✓ bestätigt" : "warte auf Bestätigung …"));
      lines.push("Abweichung von neutral: " + (v.test.departure_confirmed ? "✓ gemessen" : "wird geprüft …"));
      if (v.test.departure_confirmed) {
        lines.push("Rückkehr nach neutral: " + (v.test.settle_samples || 0) + " von " +
          (v.test.settle_required || 0) + " Messwerten in Folge");
      }
      if (v.test.live_battery_kw != null) {
        lines.push("Batterie jetzt: " + v.test.live_battery_kw.toFixed(2) + " kW");
      }
      lines.push("Verbleibende Zeit bis zum sicheren Abbruch: " + (v.test.seconds_remaining || 0) + " s");
      actions += '<button type="button" class="cal-btn ghost" data-act="abort">Test abbrechen</button>';
    } else {
      actions += '<button type="button" class="cal-btn" data-act="test">Neutral-Zeit-Test starten (' +
        v.test_kw + ' kW, max. ' + Math.round((v.test_ttl_seconds || 0) / 60) + ' min)</button>';
    }

    if (v.evidence && v.evidence.valid) {
      var ev = v.evidence;
      var line = "Letztes Ergebnis (vor " + ev.age_seconds + " s): " + verdictPill(ev.verdict);
      if (ev.verdict === "bestanden" && ev.measured_seconds != null) {
        line += " - gemessen " + ev.measured_seconds + " s";
      }
      lines.push(line);
      if (ev.reason) lines.push(ev.reason);
      if (v.can_record) {
        actions += '<button type="button" class="cal-btn primary" data-act="record">Als Nachweis übernehmen</button>';
      }
    }

    body.innerHTML =
      '<p class="cal-hint">' + lines.join("<br>") + "</p>" +
      '<div class="cal-actions">' + actions + "</div>";

    body.onclick = function (ev2) {
      var t = ev2.target;
      if (!t || !t.getAttribute) return;
      var a = t.getAttribute("data-act");
      if (a === "test") act("/api/neutral/test");
      else if (a === "abort") act("/api/neutral/abort");
      else if (a === "record") act("/api/neutral/record");
    };
  }

  function refresh() {
    return fetch("/api/neutral").then(function (r) { return r.json(); }).then(function (j) {
      render(j && j.neutral);
    }).catch(function () { /* transient - next poll */ });
  }

  if ($("neutralCard")) {
    refresh();
    global.setInterval(refresh, 3000);
  }

  // --- Teil C: der Autonomie-Schalter OHNE SHELL (ota/autonomy.json). -----
  // Er setzt AUSSCHLIESSLICH die eine Datei, die der Sidecar ohnehin jeden
  // Takt liest - jedes Tor der Torkette gilt unveraendert.
  var autonomyErr = $("autonomyErr");
  var autonomyToggle = $("autonomyToggle");

  function showAutonomyErr(msg) {
    if (!autonomyErr) return;
    autonomyErr.textContent = msg || "";
    autonomyErr.hidden = !msg;
  }

  function refreshAutonomy() {
    return fetch("/api/ota/autonomy").then(function (r) { return r.json(); }).then(function (au) {
      if (autonomyToggle) autonomyToggle.checked = !!(au && au.enabled);
    }).catch(function () { /* transient - next poll */ });
  }

  if (autonomyToggle) {
    autonomyToggle.addEventListener("change", function () {
      var enabled = autonomyToggle.checked;
      showAutonomyErr("");
      post("/api/ota/autonomy", { enabled: enabled }).then(function (resp) {
        if (ensureToken(resp)) {
          post("/api/ota/autonomy", { enabled: enabled }).then(function (retry) {
            if (!retry.ok) { showAutonomyErr((retry.body && retry.body.error) || "Aktion fehlgeschlagen."); refreshAutonomy(); }
          });
          return;
        }
        if (!resp.ok) { showAutonomyErr((resp.body && resp.body.error) || "Aktion fehlgeschlagen."); refreshAutonomy(); }
      });
    });
    refreshAutonomy();
    global.setInterval(refreshAutonomy, 5000);
  }

  global.VPNeutral = { refresh: refresh, render: render };
})(window);
